### PR TITLE
Generate OAS3 as code

### PR DESCRIPTION
**Pull request by rgreinho**
Leverages `utoipa` to generate the OAS3.

Changes:
- Splits `/ratings` and `/pipelines`.
- Moves schemas to schema.rs.
- Adds OAS 3.1 from Axum.
- Adds OAS 3.0, down-converted from the OAS 3.1.
- Regenerates the bnaclient.
- Updates the Justfile.
- Makes enums not optional.
- Uses APIErrors struct for errors.
- Adds ApiError examples for each support HTTP status (400, 401, 403, 404).
- Adds ErrorResponses enum to group all errors.

Signed-off-by: Rémy Greinhofer <remy.greinhofer@gmail.com>

### CODE REVIEW

File: lambdas/src/core/resource/cities/endpoint.rs [hunk 1]
```diff
@@ -111,45 +240,112 @@ struct SubmissionParameters {
...
+  get,
+  path = "/cities/submissions/{submission_id}",
+  description = "Get the details of a specific sumission.",
+  tag = TAG,
+  params(
+    ("submission_id" = i32, Path, description = "Submission identifier", example = "1"),
+    ("status" = Option<str>, Query, description = "Filter for the submission status", example = "Pending"),
+  ),
+  responses(
+    (status = OK, description = "Fetches a submission", body = Submissions),
```
Comment (robahtou) [inline lambdas/src/core/resource/cities/endpoint.rs:253]: should this be `Submission` or keep it pluarl? the endpoint describes a single entity.

File: lambdas/src/core/resource/cities/endpoint.rs [hunk 2]
```diff
@@ -111,45 +240,112 @@ struct SubmissionParameters {
...
+    ErrorResponses,
+  ))]
 async fn get_cities_submissions(
     Query(submission_params): Query<SubmissionParameters>,
     pagination: Option<Query<PaginationParameters>>,
-) -> Result<Json<Value>, ExecutionError> {
+) -> Result<PageFlow<Submissions>, ExecutionError> {
     let Query(pagination) = pagination.unwrap_or_default();
-    get_cities_submissions_adaptor(
+    let cities_submissions = get_cities_submissions_adaptor(
```
Comment (robahtou) [inline lambdas/src/core/resource/cities/endpoint.rs:261]: same concern as above. Should this be singular or pluaral? `get_cities_submission_adaptor` ?

File: lambdas/src/core/resource/pipelines/endpoint.rs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    get_pipelines_bna_adaptor(pipeline_id, ctx)
+        .await
+        .map(BnaPipeline::from)
+        .map(Json)
+}
+
+#[utoipa::path(
+  get,
+  path = "/pipelines/bna",
+  description = "Get the details of a specific BNA pipeline",
```
Comment (robahtou) [inline lambdas/src/core/resource/pipelines/endpoint.rs:54]: i think the description needs to be updated? this is a fetch all request?

File: lambdas/src/core/resource/ratings/endpoint.rs [hunk 4]
```diff
@@ -1,137 +1,126 @@
...
+        payload,
+    ))
 }
 
-async fn post_ratings_analysis(
-    Json(bna_pipeline): Json<BNAPipelinePost>,
-) -> Result<(StatusCode, Json<Value>), ExecutionError> {
-    post_ratings_analysis_adaptor(bna_pipeline)
+#[utoipa::path(
+  post,
```
Comment (robahtou) [inline lambdas/src/core/resource/ratings/endpoint.rs:82]: should this have a request body for the POST request?

**Comment by robahtou** [comment]
noticed that many of the endpoints have: `page_size` & `page`. Can these be it's own component schema and use a reference?

**Comment by robahtou** [comment]
going over the response errors. Can you make them their own schema? For example, 

```
components:
  responses:
    BadRequest:
      description: The request was formatted incorrectly or missing required parameters.
      content:
        application/json:
          schema:
            $ref: '#/components/schemas/APIError'
...
```
Then,
```
 /cities:
    get:
      responses:
        '400':
          $ref: '#/components/responses/BadRequest'
```
something like that?

File: lambdas/src/core/resource/cities/endpoint.rs [hunk 5]
```diff
@@ -1,108 +1,237 @@
...
+    (status = CREATED, description = "Creates a new city", body = City),
+    ErrorResponses,
+  ))]
+async fn post_city(Json(city): Json<city::CityPost>) -> Result<Json<Value>, ExecutionError> {
     post_cities_adaptor(city).await.map(Json)
 }
 
+#[utoipa::path(
+  get,
+  path = "/cities/submissions/",
```
[the remark this answers is not in the document]
Comment (rgreinho) [reply lambdas/src/core/resource/cities/endpoint.rs]: Addressed ✅

File: lambdas/src/core/resource/cities/endpoint.rs (hunk 1 above)
Comment (rgreinho) [reply lambdas/src/core/resource/cities/endpoint.rs:253]: It is singular when it returns a single result, and plural when it returns multiple ones.

Comment (rgreinho) [reply lambdas/src/core/resource/cities/endpoint.rs:253]: Oh, I see! it does not match the description!

Comment (rgreinho) [reply lambdas/src/core/resource/cities/endpoint.rs:253]: OK, it was reversed. Fixed now. Good catch!

File: lambdas/src/core/resource/pipelines/endpoint.rs (hunk 3 above)
Comment (rgreinho) [reply lambdas/src/core/resource/pipelines/endpoint.rs:54]: Yep! Good catch again!

File: lambdas/src/core/resource/cities/endpoint.rs (hunk 2 above)
Comment (rgreinho) [reply lambdas/src/core/resource/cities/endpoint.rs:261]: Hmm, looks like there is a mismatch between the endpoint returning a single results, and the endpoint returning multiple ones...

Comment (rgreinho) [reply lambdas/src/core/resource/cities/endpoint.rs:261]: Fixed ✅

File: lambdas/src/core/resource/ratings/endpoint.rs (hunk 4 above)
Comment (rgreinho) [reply lambdas/src/core/resource/ratings/endpoint.rs:82]: It totally should! I completely missed it 😱

**Comment by rgreinho** [comment]
> going over the response errors. Can you make them their own schema? For example,
> 
> ```
> components:
>   responses:
>     BadRequest:
>       description: The request was formatted incorrectly or missing required parameters.
>       content:
>         application/json:
>           schema:
>             $ref: '#/components/schemas/APIError'
> ...
> ```
> 
> Then,
> 
> ```
>  /cities:
>     get:
>       responses:
>         '400':
>           $ref: '#/components/responses/BadRequest'
> ```
> 
> something like that?

I think the generator considers them as custom objects and therefore inlines them. This is because I had to created custom examples for each one with the specific error codes and messages.

File: lambdas/src/core/resource/ratings/endpoint.rs (hunk 4 above)
Comment (rgreinho) [reply lambdas/src/core/resource/ratings/endpoint.rs:84]: Fixed ✅

**Comment by rgreinho** [comment]
> noticed that many of the endpoints have: `page_size` & `page`. Can these be it's own component schema and use a reference?

Done ✅

**Pull request merged**